### PR TITLE
[Workflow] Choose which Workflow events should be dispatched

### LINKED ISSUE #36617: [Workflow] Choose which Workflow events should be dispatched
**Description**  
Provide a way to choose which events are to be dispatched by the Workflow component.

**Background**  
Currently I have an application that processes orders. See below for Workflow places and transitions:
```
            places:
                - new
                - sent
                - saved_order_id
                - in_progress
                - order_complete
                - activated
                - complete
                - failed
            transitions:
                send:
                    from: new
                    to:   sent
                save_order_id:
                    from: sent
                    to: saved_order_id
                check:
                    from: saved_order_id
                    to: in_progress
                complete_order:
                    from: in_progress
                    to: order_complete
                activate:
                    from: order_complete
                    to: activated
                complete:
                    from: activated
                    to:   complete
                fail:
                    from: [ 'new', 'sent', 'saved_order_id', 'in_progress', 'order_complete', 'activated' ]
                    to: failed
```

I have workflows that assist in maintaining the state of the order. Currently my code listens for Guard Events and performs some action based on the state of the workflow. 

For example when attempting to place an order via a third party API I have implemented a Guard Event Listener for the `send` transition. In the EventListener I attempt to call the third party API to place the order, if this fails I return a TransitionBlocker which prevents the transition from taking place. This all works perfectly.

My current issue is that when a transition has taken place successfully, the Workflow component calls the `announce()` method so that we can find which new transitions are now possible for the new place. The issue with this is that the `announce()` method eventually calls `buildTransitionBlockerListForTransition()` but with possible transitions for the new place the order is in. In the above example if the `send` transition completes successfully `buildTransitionBlockerListForTransition()` is called with `save_order_id` as the transition.

This then dispatches the Guard Event for the `save_order_id` transition even though we’ve only explicitly called Workflow->apply() on the `send` transition. The effect of this is that Guard Event Listeners for the `save_order_id` transition fire prematurely.

While this isn’t an issue for some transitions of the Workflow, for Guard Event Listeners that contact a third party API, or perform some kind of state change with the order, it means this state change happens before we explicitly call the transition where it should occur. Then when we actually call the next transition ourselves sometime later it happens for a second time.

**Proposal** 
Provide some method of choosing which events the Workflow component should dispatch for a specific workflow.

In an ideal scenario I think this could be added as an optional configuration item for each workflow definition. An array that contains all events that should be dispatched, or the inverse, an array that defines which events should be excluded from being dispatched.

For example:

```
framework:
    workflows:
        some_workflow:
            type: state_machine
            events: ['leave', 'enter', 'entered', 'completed']
            audit_trail:
                enabled: true
...
```

Thank you for taking the time to read this proposal :)

### PR DESCRIPTION
**Pull request by lyrixx**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #36633 ; Fix #36617
| License       | MIT
| Doc PR        |

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php [hunk 1]
```diff
@@ -381,6 +383,33 @@ private function addWorkflowSection(ArrayNodeDefinition $rootNode)
                                         ->defaultValue([])
                                         ->prototype('scalar')->end()
                                     ->end()
+                                    ->variableNode('dispatched_events')
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:386]: `dispatched_events` -> `events_to_dispatch`?

File: src/Symfony/Component/Workflow/Workflow.php [hunk 2]
```diff
@@ -34,20 +34,45 @@
...
     private $name;
+    /**
+     * When `null` fire all events (the default behaviour).
+     * Setting this to an empty array `[]` means no events are dispatched (except the Guard Event).
+     * Passing an array with WorkflowEvents will allow only those events to be dispatched plus
+     * the Guard Event.
+     *
+     * @var array|string[]|null
+     */
+    private $eventsToBeDispatched = null;
```
Comment (fabpot) [inline src/Symfony/Component/Workflow/Workflow.php:67]: `EventsToDispatch`?

File: src/Symfony/Component/Workflow/Workflow.php [hunk 3]
```diff
@@ -34,20 +34,45 @@
...
+        WorkflowEvents::ENTERED => self::DISABLE_ENTERED_EVENT,
+        WorkflowEvents::COMPLETED => self::DISABLE_COMPLETED_EVENT,
+        WorkflowEvents::ANNOUNCE => self::DISABLE_ANNOUNCE_EVENT,
+    ];
+
     private $definition;
     private $markingStore;
     private $dispatcher;
     private $name;
+    /**
```
Comment (fabpot) [inline src/Symfony/Component/Workflow/Workflow.php:59]: missing blank line above

**Review by fabpot — CHANGES_REQUESTED** [review]
LGTM, with minor comments about the naming.

**Comment by lyrixx** [comment]
@fabpot thanks for the review. I have addressed your comments

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/workflow.php [hunk 4]
```diff
@@ -25,6 +25,7 @@
                 abstract_arg('marking store'),
                 service('event_dispatcher')->ignoreOnInvalid(),
                 abstract_arg('workflow name'),
+                abstract_arg('dispatched events'),
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/workflow.php:28]: events to dispatch?

File: src/Symfony/Component/Workflow/Workflow.php [hunk 5]
```diff
@@ -34,20 +34,46 @@
...
-    public function __construct(Definition $definition, MarkingStoreInterface $markingStore = null, EventDispatcherInterface $dispatcher = null, string $name = 'unnamed')
+    /**
+     * When `null` fire all events (the default behaviour).
+     * Setting this to an empty array `[]` means no events are dispatched (except the Guard Event).
+     * Passing an array with WorkflowEvents will allow only those events to be dispatched plus
+     * the Guard Event.
+     *
+     * @var array|string[]|null
+     */
+    private $eventToDispatch = null;
```
Comment (fabpot) [inline src/Symfony/Component/Workflow/Workflow.php:68]: event**s**ToDispatch?

Comment (lyrixx) [reply src/Symfony/Component/Workflow/Workflow.php]: Oups, fixed 👍🏼

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @lyrixx.

**Pull request merged**